semanticdb/symtab/src/main/scala/scala/meta/internal/symtab/GlobalSymbolTable.scala

```diff
@@ -0,0 +1,72 @@
+package scala.meta.internal.symtab
+
+import scala.annotation.tailrec
+import scala.collection.concurrent.TrieMap
+import scala.meta.internal.classpath.ClasspathIndex
+import scala.meta.internal.semanticdb.SymbolInformation
+import scala.meta.io._
+import scala.meta.internal.metacp._
+import scala.meta.internal.scalacp.Scalalib
+import scala.meta.internal.semanticdb.Scala._
+import scala.reflect.NameTransformer
+
+/** A lazy symbol table that returns global symbols on-the-fly from disk. */
+final class GlobalSymbolTable(classpathIndex: ClasspathIndex) extends SymbolTable {
+
+  private val symbolCache = TrieMap.empty[String, SymbolInformation]
+  Scalalib.synthetics.foreach(enter)
+
+  override def toString: String = s"GlobalSymbolTable($classpathIndex)"
+
+  private def enter(infos: ClassfileInfos): Unit =
+    infos.infos.foreach { info =>
+      symbolCache(info.symbol) = info
+    }
+
+  @tailrec private def getToplevel(symbol: String): String = {
+    val owner = symbol.owner
+    if (owner.isPackage) symbol
+    else getToplevel(owner)
+  }
+
+  private def loadSymbol(symbol: String): Unit = {
+    val toplevel = getToplevel(symbol)
+    val classdir = toplevel.owner
```

Comment:
`symbol.ownerChain.dropWhile(!_.isPackage).head`?

Reply:
That doesn't read as nicely IMO

Reply:
As it currently stands, `getToplevel` freezes on symbols that aren't owned by packages (e.g. locals, nosymbol and root). I'd prefer if we used a standard way to recur on owners.